include/nbl/builtin/hlsl/shapes/spherical_triangle.hlsl

```diff
@@ -0,0 +1,132 @@
+        return solidAngleOfTriangle(dummy0,dummy1,dummy2,dummy3,dummy4,dummy5);
+    }
+
+    scalar_type projectedSolidAngleOfTriangle(NBL_CONST_REF_ARG(vector3_type) receiverNormal, NBL_REF_ARG(vector3_type) cos_sides, NBL_REF_ARG(vector3_type) csc_sides, NBL_REF_ARG(vector3_type) cos_vertices)
+    {
+        if (pyramidAngles())
+            return 0.f;
+
+        vector3_type awayFromEdgePlane0 = hlsl::cross<vector3_type>(vertex1, vertex2) * csc_sides[0];
+        vector3_type awayFromEdgePlane1 = hlsl::cross<vector3_type>(vertex2, vertex0) * csc_sides[1];
+        vector3_type awayFromEdgePlane2 = hlsl::cross<vector3_type>(vertex0, vertex1) * csc_sides[2];
+
+        // useless here but could be useful somewhere else
+        cos_vertices[0] = hlsl::dot<vector3_type>(awayFromEdgePlane1, awayFromEdgePlane2);
+        cos_vertices[1] = hlsl::dot<vector3_type>(awayFromEdgePlane2, awayFromEdgePlane0);
+        cos_vertices[2] = hlsl::dot<vector3_type>(awayFromEdgePlane0, awayFromEdgePlane1);
+        // TODO: above dot products are in the wrong order, either work out which is which, or try all 6 permutations till it works
+        cos_vertices = hlsl::clamp<vector3_type>((cos_sides - cos_sides.yzx * cos_sides.zxy) * csc_sides.yzx * csc_sides.zxy, hlsl::promote<vector3_type>(-1.0), hlsl::promote<vector3_type>(1.0));
+
+        matrix<scalar_type, 3, 3> awayFromEdgePlane = matrix<scalar_type, 3, 3>(awayFromEdgePlane0, awayFromEdgePlane1, awayFromEdgePlane2);
+        const vector3_type externalProducts = hlsl::abs(hlsl::mul(/* transposed already */awayFromEdgePlane, receiverNormal));
+
+        const vector3_type pyramidAngles = acos<scalar_type>(cos_sides);
+        return hlsl::dot<vector3_type>(pyramidAngles, externalProducts) / (2.f * numbers::pi<scalar_type>);
+    }
+
+    vector3_type vertex0;
+    vector3_type vertex1;
+    vector3_type vertex2;
+    vector3_type cos_sides;
+    vector3_type csc_sides;
+};
+
+namespace util
+{
+  // Use this convetion e_i = v_{i+2}-v_{i+1}. vertex index is modulo by 3.
+  template <typename float_t>
+  vector<float_t, 3> compInternalAngle(NBL_CONST_REF_ARG(vector<float_t, 3>) e0, NBL_CONST_REF_ARG(vector<float_t, 3>) e1, NBL_CONST_REF_ARG(vector<float_t, 3>) e2)
+  {
+    // Calculate this triangle's weight for each of its three m_vertices
+    // start by calculating the lengths of its sides
+    const float_t a = hlsl::dot(e0, e0);
+    const float_t asqrt = hlsl::sqrt(a);
+    const float_t b = hlsl::dot(e1, e1);
+    const float_t bsqrt = hlsl::sqrt(b);
+    const float_t c = hlsl::dot(e2, e2);
+    const float_t csqrt = hlsl::sqrt(c);
+
+    const float_t angle0 = hlsl::acos((b + c - a) / (2.f * bsqrt * csqrt));
+    const float_t angle1 = hlsl::acos((-b + c + a) / (2.f * asqrt * csqrt));
+    const float_t angle2 = hlsl::numbers::pi<float_t> - (angle0 + angle1);
+    // use them to find the angle at each vertex
+    return vector<float_t, 3>(angle0, angle1, angle2);
+  }
+}
+
+}
```

Comment:
what is this used for? and why is it in the spherical_triangle.hlsl ?

its clearly for a cartesian/euclidean triangle thats flat!